event-dispatcher/src/main/java/io/sk8s/event/dispatcher/EventDispatchingHandler.java

```diff
@@ -16,55 +16,264 @@
+	public synchronized void onFunctionUnregistered(ResourceDeletedEvent<XFunction> event) {
 		XFunction functionResource = event.getResource();
 		String functionName = functionResource.getMetadata().getName();
-		this.kubernetesClient.extensions().jobs().withLabel("function", functionName).delete();
 		this.functions.remove(functionName);
+		this.actualNbReplicas.remove(functionName);
+
+		deployer.undeploy(functionResource);
+
+		this.kafkaConsumerMonitor.stopTracking(functionName, functionResource.getSpec().getInput());
+
+		updateWakeUpInterval();
+		this.notify();
+
 		logger.info("function deleted: " + functionName);
 	}
+
+	@EventListener
+	public void onTopicRegistered(ResourceAddedOrModifiedEvent<Topic> event) {
+		Topic topic = event.getResource();
+		String name = topic.getMetadata().getName();
+		topics.put(name, topic);
+		topicsReady.computeIfAbsent(name, i -> new CountDownLatch(1)).countDown();
+	}
+
+	@EventListener
+	public void onTopicUnregistered(ResourceDeletedEvent<Topic> event) {
+		Topic topic = event.getResource();
+		String name = topic.getMetadata().getName();
+		topics.remove(name);
+		topicsReady.remove(name);
+	}
+
+	@EventListener
+	public void onDeploymentRegistered(ResourceAddedOrModifiedEvent<Deployment> event) {
+		Deployment deployment = event.getResource();
+		String functionName = deployment.getMetadata().getName();
+		actualNbReplicas.put(functionName, deployment.getSpec().getReplicas());
+	}
+
+	@EventListener
+	public void onDeploymentUnregistered(ResourceDeletedEvent<Deployment> event) {
+		Deployment deployment = event.getResource();
+		String functionName = deployment.getMetadata().getName();
+		actualNbReplicas.remove(functionName);
+	}
+
+	@PreDestroy
+	public synchronized void tearDown() {
+		running = false;
+		this.notify();
+		kafkaConsumerMonitor.close();
+	}
+
+	private void updateWakeUpInterval() {
+		this.scalingFrequencyMs = this.functions.isEmpty() ? 0 : 100;
+	}
+
+	private class ScalingMonitor implements Runnable {
+
+		@Override
+		public void run() {
+			while (running) {
+				synchronized (EventDispatchingHandler.this) {
+
+					Map<KafkaConsumerMonitor.TopicGroup, List<KafkaConsumerMonitor.Offsets>> offsets = kafkaConsumerMonitor
+							.compute();
+					logOffsets(offsets);
+					Map<String, Long> lags = offsets.entrySet().stream()
+							.collect(Collectors.toMap(
+									e -> e.getKey().group,
+									e -> e.getValue().stream()
+											.mapToLong(KafkaConsumerMonitor.Offsets::getLag)
+											.max().getAsLong()));
+					functions.values().stream().forEach(
+							f -> {
+								int desired = computeDesiredNbReplicas(lags, f);
+								String name = f.getMetadata().getName();
+								Integer currentDesired = actualNbReplicas.computeIfAbsent(name, k -> desired);
+								System.out.println("Want " + desired + " for " + name);
+								if (desired < currentDesired) {
+									if (edgeInfos.computeIfAbsent(name,
```

Comment:
"edgeInfos" should be "downEdgeLingerPeriods" (or change that constant name to match something else, perhaps "scaleDownDelays"?)